File: cli/src/merkle.ts
/**
 * merkle.ts — Off-chain incremental Merkle tree for Chameleon.
 *
 * Reconstructs the on-chain Merkle tree from deposit events, computes
 * authentication paths for withdrawal proofs, and mirrors the on-chain
 * `merkle.rs` insert logic.
 *
 * Hash function: Poseidon2([left, right]) via poseidon.ts — MUST match the
 * Soroban contract and Noir circuit (verified by `make test-crypto`).
 *
 * Tree parameters:
 *   TREE_DEPTH = 20, ROOT_HISTORY_SIZE = 30
 *   Zero leaf: Field(0) (empty slot sentinel)
 *   zeros[i+1] = Poseidon2([zeros[i], zeros[i]])
 */

import { hash2, hexToField, fieldToHex } from './poseidon';

export const TREE_DEPTH = 20;
export const ROOT_HISTORY_SIZE = 30;

/**
 * Precomputed zero values at each level.
 * zeros[0] = 0 (empty leaf)
 * zeros[i+1] = Poseidon2([zeros[i], zeros[i]])
 */
export function computeZeros(): bigint[] {
  const zeros: bigint[] = [0n];
  for (let i = 0; i < TREE_DEPTH; i++) {
    zeros.push(hash2(zeros[i], zeros[i]));
  }
  return zeros;
}

/** Cached zero values (computed once). */
export const ZEROS = computeZeros();

/** Merkle path proof for a single leaf. */
export interface MerkleProof {
  /** Leaf index in the tree (= deposit index). */
  leafIndex: number;
  /** Sibling hashes along the path from leaf to root (depth=20 elements). */
  pathElements: bigint[];
  /** Direction bits: 0 = leaf/node is left child, 1 = right child. */
  pathIndices: number[];
  /** Recomputed root (should match on-chain root at deposit time). */
  root: bigint;
}

/**
 * In-memory incremental Merkle tree that mirrors the Soroban contract's state.
 *
 * Rebuilt from deposit events by inserting commitments in order.
 */
export class MerkleTree {
  private leaves: bigint[] = [];
  private filledSubtrees: bigint[];
  private roots: bigint[] = [];
  private nextIndex = 0;

  constructor(readonly depth = TREE_DEPTH) {
    this.filledSubtrees = ZEROS.slice(0, depth);
  }

  /**
   * Insert a leaf (commitment) into the tree, updating filled subtrees and root.
   *
   * Mirrors `merkle.rs::insert()`.
   *
   * @param leaf 32-byte commitment as a BigInt field element
   * @returns The new Merkle root after insertion
   */
  insert(leaf: bigint): bigint {
    this.leaves.push(leaf);
    const idx = this.nextIndex++;
    let current = leaf;
    let filledIdx = idx;

    for (let i = 0; i < this.depth; i++) {
      let left: bigint, right: bigint;
      if (filledIdx % 2 === 0) {
        // Left child: store as new filled subtree, pair with zero
        left = current;
        right = ZEROS[i];
        this.filledSubtrees[i] = current;
      } else {
        // Right child: pair with stored left subtree
        left = this.filledSubtrees[i];
        right = current;
      }
      current = hash2(left, right);
      filledIdx = Math.floor(filledIdx / 2);
    }

    this.roots.push(current);
    return current;
  }

  /**
   * Compute the Merkle authentication path for the leaf at `leafIndex`.
   *
   * Rebuilds the full tree from all stored leaves to compute accurate siblings.
   *
   * @param leafIndex The position of the target leaf
   * @returns MerkleProof with pathElements, pathIndices, and root
   * @throws If leafIndex is out of range
   */
  generateProof(leafIndex: number): MerkleProof {
    if (leafIndex >= this.leaves.length) {
      throw new Error(`Leaf index ${leafIndex} out of range (tree has ${this.leaves.length} leaves)`);
    }

    // Build full tree level by level
    const treeSize = Math.pow(2, this.depth);
    // Level 0 = leaves
    const levels: bigint[][] = [];
    const leafLevel = [...this.leaves];
    // Pad with zeros
    while (leafLevel.length < treeSize) leafLevel.push(ZEROS[0]);
    levels.push(leafLevel);

    // Build upward
    for (let i = 0; i < this.depth; i++) {
      const prev = levels[i];
      const next: bigint[] = [];
      for (let j = 0; j < prev.length; j += 2) {
        next.push(hash2(prev[j], prev[j + 1] ?? ZEROS[i]));
      }
      levels.push(next);
    }

    // Root is levels[depth][0]
    const root = levels[this.depth][0];

    // Collect path elements and indices
    const pathElements: bigint[] = [];
    const pathIndices: number[] = [];
    let idx = leafIndex;

    for (let i = 0; i < this.depth; i++) {
      const isRight = idx % 2;
      const siblingIdx = isRight ? idx - 1 : idx + 1;
      pathElements.push(levels[i][siblingIdx] ?? ZEROS[i]);
      pathIndices.push(isRight);
      idx = Math.floor(idx / 2);
    }

    return { leafIndex, pathElements, pathIndices, root };
  }

  /** Return the current root. */
  getRoot(): bigint {
    return this.roots[this.roots.length - 1] ?? ZEROS[this.depth];
  }

  /** Return all roots (for root history ring buffer check). */
  getRootHistory(): bigint[] {
    return this.roots.slice(-ROOT_HISTORY_SIZE);
  }

  /** Return true if root is in the last ROOT_HISTORY_SIZE roots. */
  isKnownRoot(root: bigint): boolean {
    return this.getRootHistory().some(r => r === root);
  }

  /** Return the number of leaves inserted. */
  get size(): number {
    return this.leaves.length;
  }

  /**
   * Find the leaf index for a given commitment, or -1 if not found.
   */
  findLeafIndex(commitment: bigint): number {
    return this.leaves.indexOf(commitment);
  }

  /**
   * Serialize tree state for caching to disk.
   */
  serialize(): TreeState {
    return {
      leaves: this.leaves.map(l => fieldToHex(l)),
      filledSubtrees: this.filledSubtrees.map(f => fieldToHex(f)),
      roots: this.roots.map(r => fieldToHex(r)),
      nextIndex: this.nextIndex,
      depth: this.depth,
    };
  }

  /**
   * Restore tree state from a serialized snapshot.
   */
  static deserialize(state: TreeState): MerkleTree {
    const tree = new MerkleTree(state.depth);
    tree.leaves = state.leaves.map(hexToField);
    tree.filledSubtrees = state.filledSubtrees.map(hexToField);
    tree.roots = state.roots.map(hexToField);
    tree.nextIndex = state.nextIndex;
    return tree;
  }
}

/** Serializable tree state for caching. */
export interface TreeState {
  leaves: string[];       // hex-encoded field elements
  filledSubtrees: string[];
  roots: string[];
  nextIndex: number;
  depth: number;
}

/**
 * Verify a Merkle proof (matches circuit logic in merkle.nr).
 *
 * @param leaf The leaf value
 * @param pathElements Sibling hashes
 * @param pathIndices Direction bits (0=left, 1=right)
 * @param expectedRoot Expected root
 * @returns true if computed root matches expectedRoot
 */
export function verifyMerkleProof(
  leaf: bigint,
  pathElements: bigint[],
  pathIndices: number[],
  expectedRoot: bigint,
): boolean {
  let current = leaf;
  for (let i = 0; i < pathElements.length; i++) {
    const sibling = pathElements[i];
    const isRight = pathIndices[i];
    const left = isRight ? sibling : current;
    const right = isRight ? current : sibling;
    current = hash2(left, right);
  }
  return current === expectedRoot;
}

File: contracts/privacy_pool/src/merkle.rs
//! Incremental Merkle tree implementation (Tornado pattern) for the PrivacyPool.
//!
//! Uses Poseidon2 over BN254 scalar field (via `soroban-poseidon`) for hashing.
//! The tree is depth 20, supporting up to 2^20 = 1,048,576 leaves.
//!
//! State stored on-chain:
//! - `filled_subtrees[i]` — the "frontier" node at level `i`
//! - Root history ring buffer (last `ROOT_HISTORY_SIZE` roots)
//!
//! IMPORTANT: The Poseidon2 parameters here (t=4, BnScalar) MUST match those used
//! in the Noir circuit (`std::hash::poseidon2`) and the TypeScript SDK.
//! Cross-implementation test (`make test-crypto`) verifies this.

extern crate alloc;
use alloc::vec::Vec;

use soroban_poseidon::{poseidon2_hash, Field};
use soroban_sdk::{crypto::BnScalar, Bytes, BytesN, Env, U256, Vec as SorobanVec};

use crate::storage::{
    get_filled_subtree, get_next_index, get_root_history_current, set_filled_subtree,
    set_next_index, set_root, set_root_history, set_root_history_current,
};

/// Merkle tree depth. Supports 2^20 = 1,048,576 leaves.
pub const TREE_DEPTH: u32 = 20;

/// Maximum number of leaves.
pub const MAX_LEAVES: u32 = 1u32 << TREE_DEPTH;

/// Number of historical roots to store (ring buffer). Provers have this many
/// deposits of "breathing room" before their root becomes stale.
pub const ROOT_HISTORY_SIZE: u32 = 30;

/// Compute Poseidon2(left, right) → 32-byte hash.
///
/// Uses `soroban-poseidon` with `t=4` (absorption width) over the BN254 scalar field.
/// Inputs are treated as field elements (32-byte big-endian, reduced mod p).
pub fn poseidon2_pair(env: &Env, left: &BytesN<32>, right: &BytesN<32>) -> BytesN<32> {
    let modulus = <BnScalar as Field>::modulus(env);
    let la = Bytes::from_array(env, &left.to_array());
    let ra = Bytes::from_array(env, &right.to_array());
    let mut inputs: SorobanVec<U256> = SorobanVec::new(env);
    inputs.push_back(U256::from_be_bytes(env, &la).rem_euclid(&modulus));
    inputs.push_back(U256::from_be_bytes(env, &ra).rem_euclid(&modulus));
    let out = poseidon2_hash::<4, BnScalar>(env, &inputs);
    let out_bytes = out.to_be_bytes();
    let mut arr = [0u8; 32];
    out_bytes.copy_into_slice(&mut arr);
    BytesN::from_array(env, &arr)
}

/// Compute Poseidon2([input]) → 32-byte hash (single-element absorption).
///
/// Used for nullifier hash: `NH = Poseidon2([nullifier])`.
pub fn poseidon2_one(env: &Env, input: &BytesN<32>) -> BytesN<32> {
    let modulus = <BnScalar as Field>::modulus(env);
    let ia = Bytes::from_array(env, &input.to_array());
    let mut inputs: SorobanVec<U256> = SorobanVec::new(env);
    inputs.push_back(U256::from_be_bytes(env, &ia).rem_euclid(&modulus));
    let out = poseidon2_hash::<4, BnScalar>(env, &inputs);
    let out_bytes = out.to_be_bytes();
    let mut arr = [0u8; 32];
    out_bytes.copy_into_slice(&mut arr);
    BytesN::from_array(env, &arr)
}

/// Precompute the zero values for all levels of the Merkle tree.
///
/// `zeros[0]` = `[0u8; 32]` (empty leaf sentinel)
/// `zeros[i+1]` = `Poseidon2(zeros[i], zeros[i])`
///
/// These are deterministic and used during `insert` to fill unpopulated siblings.
/// Called once during `initialize`.
pub fn compute_zeros(env: &Env) -> Vec<BytesN<32>> {
    let mut zeros: Vec<BytesN<32>> = Vec::with_capacity((TREE_DEPTH + 1) as usize);
    let mut cur = BytesN::from_array(env, &[0u8; 32]);
    zeros.push(cur.clone());
    for _ in 0..TREE_DEPTH {
        cur = poseidon2_pair(env, &cur, &cur);
        zeros.push(cur.clone());
    }
    zeros
}

/// Initialize the incremental Merkle tree.
///
/// Computes and stores the initial `filled_subtrees` (all equal to `zeros[level]`)
/// and the initial root (`zeros[TREE_DEPTH]`).
/// Also writes the initial root into ring-buffer position 0.
///
/// Must be called exactly once during contract `initialize`.
pub fn init_tree(env: &Env) {
    let zeros = compute_zeros(env);
    for level in 0..TREE_DEPTH {
        set_filled_subtree(env, level, &zeros[level as usize]);
    }
    let initial_root = &zeros[TREE_DEPTH as usize];
    set_root(env, initial_root);
    set_root_history(env, 0, initial_root);
    set_next_index(env, 0);
    set_root_history_current(env, 0);
}

/// Insert a new leaf into the Merkle tree and update roots.
///
/// Implements the incremental update from Tornado Cash:
/// - Traverses from leaf to root, replacing the filled subtree at each level.
/// - Emits the new root into the ring buffer.
///
/// # Arguments
/// - `env`    — Soroban environment
/// - `leaf`   — 32-byte commitment to insert
///
/// # Returns
/// - `(new_root, leaf_index)` where `leaf_index` is the position of the inserted leaf
///
/// # Errors
/// Returns `(zero_root, u32::MAX)` if the tree is full — caller must check next_index before calling.
pub fn insert(env: &Env, leaf: &BytesN<32>) -> (BytesN<32>, u32) {
    let zeros = compute_zeros(env);
    let current_index = get_next_index(env);

    let mut current_level_hash = leaf.clone();
    let mut left;
    let mut right;
    let mut filled_index = current_index;

    for i in 0..TREE_DEPTH {
        if filled_index % 2 == 0 {
            // left node: store current as the new filled subtree, pair with zero
            left = current_level_hash.clone();
            right = zeros[i as usize].clone();
            set_filled_subtree(env, i, &current_level_hash);
        } else {
            // right node: pair with the stored filled subtree
            left = get_filled_subtree(env, i).unwrap_or_else(|| zeros[i as usize].clone());
            right = current_level_hash.clone();
        }
        current_level_hash = poseidon2_pair(env, &left, &right);
        filled_index /= 2;
    }

    // Update root and root history
    let new_root = current_level_hash;
    set_root(env, &new_root);

    let hist_pos = get_root_history_current(env);
    let next_hist_pos = (hist_pos + 1) % ROOT_HISTORY_SIZE;
    set_root_history(env, next_hist_pos, &new_root);
    set_root_history_current(env, next_hist_pos);

    set_next_index(env, current_index + 1);

    (new_root, current_index)
}

/// Check if a root is in the ring buffer of recently accepted roots.
///
/// Returns `true` if `root` matches any of the last `ROOT_HISTORY_SIZE` roots.
/// The initial root is stored at position 0 during `init_tree`.
pub fn is_known_root(env: &Env, root: &BytesN<32>) -> bool {
    // Fast path: check current root
    if let Some(current) = crate::storage::get_root(env) {
        if &current == root {
            return true;
        }
    }
    // Scan ring buffer
    for i in 0..ROOT_HISTORY_SIZE {
        if let Some(h) = crate::storage::get_root_history(env, i) {
            if &h == root {
                return true;
            }
        }
    }
    false
}
